Remove skip_if_not_multigpu decorator

H-Huang opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/54887

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54916 Remove skip_if_not_multigpu decorator**

`@skip_if_not_multigpu` was skipping all the tests that use it. Removed it and replaced all references with `@skip_if_lt_x_gpu(2)` which performs the same check correctly.

Differential Revision: [D27412193](https://our.internmc.facebook.com/intern/diff/D27412193)

H-Huang committed:
Fix skip_if_not_multigpu decorator

[ghstack-poisoned]

H-Huang committed:
Update on "Fix skip_if_not_multigpu decorator"

Fixes https://github.com/pytorch/pytorch/issues/54887


`@skip_if_not_multigpu` was skipping all the tests that use it.

Differential Revision: [D27412193](https://our.internmc.facebook.com/intern/diff/D27412193)

[ghstack-poisoned]

H-Huang committed:
Update on "Fix skip_if_not_multigpu decorator"

Fixes https://github.com/pytorch/pytorch/issues/54887


`@skip_if_not_multigpu` was skipping all the tests that use it.

Differential Revision: [D27412193](https://our.internmc.facebook.com/intern/diff/D27412193)

[ghstack-poisoned]

rohan-varma commented on test/quantization/test_quantize_fx.py:
```diff
@@ -3801,9 +3799,11 @@ def print_diffs(diffs):
     def test_resnet18_ddp(self):
+        # TODO: This test is always failings and needs to be updated
+        # Issue: https://github.com/pytorch/pytorch/issues/54979
+        raise unittest.SkipTest()
```
nit: there is a decorator @unittest.skipTest that always skips the test

rohan-varma reviewed (approved):
Thanks for fixing!

gcramer23 commented:
looks good. I noticed the problem is that the func argument in def skip_if_not_multigpu. If you remove the argument it works as intended. also matches the pattern of the other defs with ```    def decorator(func):
        @wraps(func)```

gcramer23 commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -68,17 +68,14 @@ def wrapper(*args, **kwargs):
 
 def skip_if_not_multigpu(func):
```
looks good. I noticed the problem is that the func argument in def skip_if_not_multigpu. If you remove the argument it works as intended. also matches the pattern of the other defs with def decorator(func): @wraps(func)

pmeier replied:
The problem is that `skip_if_not_multigpu` used the double wrapper structure which is meant for parametrizable decorators. Removing one wrapper would have been sufficient. See my proposal in #53502.

mrshenli commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -68,17 +68,14 @@ def wrapper(*args, **kwargs):
-    return decorator
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        if torch.cuda.is_available() and torch.cuda.device_count() >= 2:
```
is this the same as `@skip_if_lt_x_gpu(2)`? If so, will it work if we consolidate all `@skip_if_not_multigpu` occurrences into  `@skip_if_lt_x_gpu(2)`?

H-Huang replied:
Yep, great point. I was working on just removing this entirely and replacing all references to `@skip_if_not_multigpu` with `skip_if_lit_x_gpu(2)`, just need to figure out the failures for the tests that are now reenabled.

H-Huang committed:
Update on "Fix skip_if_not_multigpu decorator"

Fixes https://github.com/pytorch/pytorch/issues/54887


`@skip_if_not_multigpu` was skipping all the tests that use it.

Differential Revision: [D27412193](https://our.internmc.facebook.com/intern/diff/D27412193)

[ghstack-poisoned]

H-Huang committed:
Update on "Remove skip_if_not_multigpu decorator"

Fixes https://github.com/pytorch/pytorch/issues/54887


`@skip_if_not_multigpu` was skipping all the tests that use it. Removed it and replaced all references with `@skip_if_lt_x_gpu(2)` which performs the same check correctly.

Differential Revision: [D27412193](https://our.internmc.facebook.com/intern/diff/D27412193)

[ghstack-poisoned]

mrshenli reviewed (approved):
Thanks for fixing, LGTM! Let's wait for all tests. :)

mrshenli commented:
One caveat is that, PR CIs won't run multi-gpu tests (4 GPUs), and that only runs on master or ci-all PRs. Is it necessary to try ci-all on this change?

H-Huang commented:
Thanks for the tip @mrshenli! I will check that tests pass for https://github.com/pytorch/pytorch/pull/55084 before merging

zhaojuanmao commented:
I'm wondering when those tests are disabled...?

H-Huang committed:
Update on "Remove skip_if_not_multigpu decorator"

Fixes https://github.com/pytorch/pytorch/issues/54887


`@skip_if_not_multigpu` was skipping all the tests that use it. Removed it and replaced all references with `@skip_if_lt_x_gpu(2)` which performs the same check correctly.

Differential Revision: [D27412193](https://our.internmc.facebook.com/intern/diff/D27412193)

[ghstack-poisoned]

H-Huang commented:
test failure in rocm, looks like it should be skipped so adding @skip_if_rocm to `test_grad_layout_1devicemodule_1replicaperprocess`
